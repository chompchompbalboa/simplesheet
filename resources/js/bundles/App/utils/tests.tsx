//-----------------------------------------------------------------------------
// Imports
//-----------------------------------------------------------------------------
import React from 'react'
import { Provider as ReduxProvider } from 'react-redux'
import { applyMiddleware, createStore } from 'redux'
import thunkMiddleware from 'redux-thunk'
import { render } from '@testing-library/react'

import { appReducer } from '@app/state'

//-----------------------------------------------------------------------------
// With Redux
//-----------------------------------------------------------------------------
// @ts-ignore
const WithRedux = ({ children }) => {
  const store = createStore(appReducer, applyMiddleware(thunkMiddleware))
  return (
    <ReduxProvider store={store}>
      { children }
    </ReduxProvider>
  )
}

//-----------------------------------------------------------------------------
// Exports
//-----------------------------------------------------------------------------
export * from '@testing-library/react'
// @ts-ignore
export const renderWithRedux = (ui, options?) => render(ui, { wrapper: WithRedux, ...options })
