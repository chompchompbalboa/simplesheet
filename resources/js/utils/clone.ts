const clone = (itemToClone: any): any => {
	return JSON.parse(JSON.stringify(itemToClone))
}

export default clone
