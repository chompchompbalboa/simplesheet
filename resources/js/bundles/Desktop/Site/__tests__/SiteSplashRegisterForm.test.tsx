//-----------------------------------------------------------------------------
// Imports
//-----------------------------------------------------------------------------
import React from 'react'
import 'jest-styled-components'
import '@testing-library/jest-dom/extend-expect'
import axiosMock from 'axios'

import { fireEvent, renderWithRedux } from '@/testing/library'
import { mockEnvironment } from '@/testing/mocks'
import { appStateFactory, IAppStateFactoryInput } from '@/testing/mocks/appState'
import { flushPromises } from '@/testing/utils'

import { 
  SiteSplashRegisterForm,
  siteSplashRegisterFormStatusMessages
} from '@desktop/Site/SiteSplashRegisterForm'

//-----------------------------------------------------------------------------
// Setup
//-----------------------------------------------------------------------------
const {
  allFiles,
  allFileIds,
  allSheetsFromDatabase
} = appStateFactory({} as IAppStateFactoryInput)

const fileId = allFileIds[0]
const sheetId = allFiles[fileId].typeId
const sheetFromDatabase = allSheetsFromDatabase[sheetId]

// @ts-ignore
axiosMock.get.mockResolvedValue({ data: sheetFromDatabase })

console.warn = jest.fn()
console.error = jest.fn()

//-----------------------------------------------------------------------------
// Tests
//-----------------------------------------------------------------------------
describe('SiteSplashRegisterForm', () => {

  const { location } = window

  beforeAll(() => {
    // @ts-ignore
    global.environment = mockEnvironment
    jest.useFakeTimers()
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  beforeEach(() => {
    delete window.location
    // @ts-ignore
    window.location = { reload: jest.fn() }
  })

  afterEach(() => {
    window.location = location
    jest.clearAllMocks()
  })

  const siteSplashRegisterForm = () => {
    const { 
      getByLabelText,
      getByPlaceholderText, 
      getByTestId,
      getByText
    } = renderWithRedux(<SiteSplashRegisterForm />)
    const nameInput = getByPlaceholderText('Name') as HTMLInputElement
    const emailInput = getByPlaceholderText('Email') as HTMLInputElement
    const passwordInput = getByPlaceholderText('Password') as HTMLInputElement
    const checkboxInput = getByLabelText('I agree to start a 30-day free trial of Tasksheet (no credit card required)') as HTMLInputElement
    const submitButton = getByText('Sign Up')
    const statusContainer = getByTestId('SiteSplashRegisterFormStatus')
    return {
      nameInput,
      emailInput,
      passwordInput,
      checkboxInput,
      submitButton,
      statusContainer
    }
  }
  
  it("renders without crashing", async () => {
    const container = renderWithRedux(<SiteSplashRegisterForm />)
    expect(container).toMatchSnapshot()
  })
  
  it("displays a 'Name' input", async () => {
    const { nameInput } = siteSplashRegisterForm()
    expect(nameInput).toBeTruthy()
  })
  
  it("displays an 'Email' input", async () => {
    const { emailInput } = siteSplashRegisterForm()
    expect(emailInput).toBeTruthy()
  })
  
  it("displays a 'Password' input", async () => {
    const { passwordInput } = siteSplashRegisterForm()
    expect(passwordInput).toBeTruthy()
  })
  
  it("displays a checkbox where the user agrees to start a 30-day trial", async () => {
    const { checkboxInput } = siteSplashRegisterForm()
    expect(checkboxInput).toBeTruthy()
  })
  
  it("changes the 'Name' value on user input", async () => {
    const { nameInput } = siteSplashRegisterForm()
    const nextNameInputValue = 'First Name'
    expect(nameInput.value).toEqual('')
    fireEvent.change(nameInput, { target: { value: nextNameInputValue }})
    expect(nameInput.value).toEqual(nextNameInputValue)
  })
  
  it("changes the 'Email' value on user input", async () => {
    const { emailInput } = siteSplashRegisterForm()
    const nextEmailInputValue = 'test@test.com'
    expect(emailInput.value).toEqual('')
    fireEvent.change(emailInput, { target: { value: nextEmailInputValue }})
    expect(emailInput.value).toEqual(nextEmailInputValue)
  })
  
  it("changes the 'Password' value on user input", async () => {
    const { passwordInput } = siteSplashRegisterForm()
    const nextPasswordInputValue = 'Super Secret Password'
    expect(passwordInput.value).toEqual('')
    fireEvent.change(passwordInput, { target: { value: nextPasswordInputValue }})
    expect(passwordInput.value).toEqual(nextPasswordInputValue)
  })
  
  it("changes the checkbox value when the user clicks", async () => {
    const { checkboxInput } = siteSplashRegisterForm()
    expect(checkboxInput.checked).toBe(false)
    fireEvent.click(checkboxInput)
    expect(checkboxInput.checked).toBe(true)
    fireEvent.click(checkboxInput)
    expect(checkboxInput.checked).toBe(false)
  })
  
  it("changes the border color of the 'Email' input to 'red' if the user enters a value that is not an email and then blurs the input", async () => {
    const { emailInput } = siteSplashRegisterForm()
    expect(emailInput.value).toBe('')
    expect(emailInput).toHaveStyleRule('border', '1px solid rgb(150,150,150)')
    fireEvent.change(emailInput, { target: { value: 'Not an email address' }})
    fireEvent.blur(emailInput)
    expect(emailInput).toHaveStyleRule('border', '1px solid red')
  })
  
  it("changes the value of the submit button to 'Signing Up...' when the user submits the form", async () => {
    const { submitButton } = siteSplashRegisterForm()
    submitButton.click()
    expect(submitButton.textContent).toBe('Signing Up...')
  })
  
  it("displays an error message when the user tries to submit the form without all the fields completed", async () => {
    const { statusContainer, submitButton } = siteSplashRegisterForm()
    submitButton.click()
    jest.advanceTimersByTime(500)
    expect(statusContainer.textContent).toBe(siteSplashRegisterFormStatusMessages['NOT_ALL_FIELDS_ARE_COMPLETE'])
  })
  
  it("displays an error message when the user tries to submit the form without a valid email", async () => {
    const { nameInput, emailInput, passwordInput, statusContainer, submitButton } = siteSplashRegisterForm()
    fireEvent.change(nameInput, { target: { value: 'Name' }})
    fireEvent.change(emailInput, { target: { value: 'Not an email' }})
    fireEvent.change(passwordInput, { target: { value: 'Password' }})
    submitButton.click()
    jest.advanceTimersByTime(500)
    expect(statusContainer.textContent).toBe(siteSplashRegisterFormStatusMessages['NOT_VALID_EMAIL'])
  })
  
  it("displays an error message when the user tries to submit the form without checking the checkbox", async () => {
    const { nameInput, emailInput, passwordInput, statusContainer, submitButton } = siteSplashRegisterForm()
    fireEvent.change(nameInput, { target: { value: 'Name' }})
    fireEvent.change(emailInput, { target: { value: 'test@test.com' }})
    fireEvent.change(passwordInput, { target: { value: 'Password' }})
    submitButton.click()
    jest.advanceTimersByTime(500)
    expect(statusContainer.textContent).toBe(siteSplashRegisterFormStatusMessages['START_TRIAL_CHECKBOX_NOT_CHECKED'])
  })
  
  it("attempts to register the user when all fields are completed, the email is valid and the checkbox is checked", async () => {
    const { nameInput, emailInput, passwordInput, checkboxInput, submitButton, statusContainer } = siteSplashRegisterForm()
    const name = 'Name'
    const email = 'test@test.com'
    const password = 'Password'
    fireEvent.change(nameInput, { target: { value: name }})
    fireEvent.change(emailInput, { target: { value: email }})
    fireEvent.change(passwordInput, { target: { value: password }})
    fireEvent.click(checkboxInput)
    submitButton.click()
    expect(statusContainer.textContent).toBe(siteSplashRegisterFormStatusMessages['REGISTERING'])
    expect(axiosMock.post).toHaveBeenCalledTimes(1)
    expect(axiosMock.post).toHaveBeenCalledWith('/user/register', { name: name, email: email, password: password })
  })
  
  it("displays an error message when the email address is already in use", async () => {
    (axiosMock.post as jest.Mock).mockRejectedValueOnce({ response: { status: 409 } })
    const { nameInput, emailInput, passwordInput, checkboxInput, submitButton, statusContainer } = siteSplashRegisterForm()
    expect(window.location.reload).not.toHaveBeenCalled()
    const name = 'Name'
    const email = 'test@test.com'
    const password = 'Password'
    fireEvent.change(nameInput, { target: { value: name }})
    fireEvent.change(emailInput, { target: { value: email }})
    fireEvent.change(passwordInput, { target: { value: password }})
    fireEvent.click(checkboxInput)
    submitButton.click()
    await flushPromises()
    expect(window.location.reload).not.toHaveBeenCalled()
    jest.advanceTimersByTime(500)
    expect(statusContainer.textContent).toBe(siteSplashRegisterFormStatusMessages['EMAIL_ALREADY_IN_USE'])
  })
  
  it("displays an error message when the registration is not successful", async () => {
    (axiosMock.post as jest.Mock).mockRejectedValueOnce({ response: { status: 500 } })
    const { nameInput, emailInput, passwordInput, checkboxInput, submitButton, statusContainer } = siteSplashRegisterForm()
    expect(window.location.reload).not.toHaveBeenCalled()
    const name = 'Name'
    const email = 'test@test.com'
    const password = 'Password'
    fireEvent.change(nameInput, { target: { value: name }})
    fireEvent.change(emailInput, { target: { value: email }})
    fireEvent.change(passwordInput, { target: { value: password }})
    fireEvent.click(checkboxInput)
    submitButton.click()
    await flushPromises()
    expect(window.location.reload).not.toHaveBeenCalled()
    jest.advanceTimersByTime(500)
    expect(statusContainer.textContent).toBe(siteSplashRegisterFormStatusMessages['ERROR_DURING_REGISTRATION'])
  })
  
  it("reloads the page when the registration is successful", async () => {    
    (axiosMock.post as jest.Mock).mockResolvedValueOnce({})
    const { nameInput, emailInput, passwordInput, checkboxInput, submitButton, statusContainer } = siteSplashRegisterForm()
    expect(window.location.reload).not.toHaveBeenCalled()
    const name = 'Name'
    const email = 'test@test.com'
    const password = 'Password'
    fireEvent.change(nameInput, { target: { value: name }})
    fireEvent.change(emailInput, { target: { value: email }})
    fireEvent.change(passwordInput, { target: { value: password }})
    fireEvent.click(checkboxInput)
    submitButton.click()
    expect(statusContainer.textContent).toBe(siteSplashRegisterFormStatusMessages['REGISTERING'])
    await flushPromises()
    expect(axiosMock.post).toHaveBeenCalledTimes(1)
    expect(axiosMock.post).toHaveBeenCalledWith('/user/register', { name: name, email: email, password: password })
    expect(window.location.reload).toHaveBeenCalled()
  })

})
