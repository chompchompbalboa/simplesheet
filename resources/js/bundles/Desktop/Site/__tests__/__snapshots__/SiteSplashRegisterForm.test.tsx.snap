// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteSplashRegisterForm renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c7 {
  width: 100%;
  margin-top: 0.5rem;
  margin-left: 0;
  cursor: pointer;
  padding: 0.5rem 1.25rem;
  border: 1px solid rgb(150,150,150);
  border-radius: 5px;
  font-size: 0.9rem;
  background-color: rgba(220,220,220,1);
  color: black;
  outline: none;
  -webkit-transition: background-color 0.1s;
  transition: background-color 0.1s;
}

.c7:hover {
  background-color: white;
  color: black;
}

.c4 {
  padding: 0.25rem 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  margin-left: 0.5rem;
  width: 100%;
}

.c5 {
  margin: 0.375rem;
  margin-left: 0;
  padding: 0.5rem 0.25rem;
  border-radius: 4px;
  font-size: 0.9rem;
}

.c1 {
  margin-left: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  width: 100%;
}

.c3 {
  width: 100%;
  margin: 0.375rem;
  padding: 0.5rem 0.25rem;
  border: none;
  border: 1px solid rgb(150,150,150);
  border-radius: 4px;
  outline: none;
  font-size: 0.9rem;
}

.c8 {
  margin-top: 0.5rem;
  text-align: left;
  color: rgb(175,0,0);
}

.c0 {
  width: 100%;
}

@media (max-width:480px) {
  .c7 {
    width: 100%;
    margin: 0.375rem;
  }
}

@media (max-width:480px) {
  .c5 {
    width: 100%;
    margin: 0.375rem 0;
  }
}

@media (max-width:480px) {
  .c3 {
    width: 100%;
    margin: 0.375rem 0;
  }
}

<body>
    <div>
      <form
        class="c0"
      >
        <div
          class="c1"
        >
          <label
            class="c2"
          >
            Name
            :
          </label>
          <input
            class="c3"
            placeholder="Name"
            type="text"
            value=""
          />
        </div>
        <div
          class="c1"
        >
          <label
            class="c2"
          >
            Email
            :
          </label>
          <input
            class="c3"
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          class="c1"
        >
          <label
            class="c2"
          >
            Password
            :
          </label>
          <input
            class="c3"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="c4"
        >
          <input
            class="c5"
            id="SiteFormCheckbox"
            type="checkbox"
          />
          <label
            class="c6"
            for="SiteFormCheckbox"
          >
            I agree to start a 30-day free trial of Tasksheet (no credit card required)
          </label>
        </div>
        <button
          class="c7"
        >
          Sign Up
        </button>
        <div
          class="c8"
          data-testid="SiteSplashRegisterFormStatus"
        />
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="sc-AykKK fRCMsB"
    >
      <div
        class="sc-AykKG kgqMhc"
      >
        <label
          class="sc-AykKH bXEuhu"
        >
          Name
          :
        </label>
        <input
          class="sc-AykKI eNDOHp"
          placeholder="Name"
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-AykKG kgqMhc"
      >
        <label
          class="sc-AykKH bXEuhu"
        >
          Email
          :
        </label>
        <input
          class="sc-AykKI eNDOHp"
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div
        class="sc-AykKG kgqMhc"
      >
        <label
          class="sc-AykKH bXEuhu"
        >
          Password
          :
        </label>
        <input
          class="sc-AykKI eNDOHp"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <div
        class="sc-AykKD IcsMq"
      >
        <input
          class="sc-AykKF hoDHRU"
          id="SiteFormCheckbox"
          type="checkbox"
        />
        <label
          class="sc-AykKE eCehFn"
          for="SiteFormCheckbox"
        >
          I agree to start a 30-day free trial of Tasksheet (no credit card required)
        </label>
      </div>
      <button
        class="sc-AykKC ilFvXC"
      >
        Sign Up
      </button>
      <div
        class="sc-AykKJ jJgrwI"
        data-testid="SiteSplashRegisterFormStatus"
      />
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
  "unmount": [Function],
}
`;
